libraries/app/application.cpp

```diff
@@ -1095,13 +1095,17 @@ void application::shutdown_plugins()
       entry.second->plugin_shutdown();
    return;
 }
-void application::shutdown(bool do_flush)
+void application::shutdown(bool after_exception)
 {
    if( my->_p2p_network )
       my->_p2p_network->close();
    if( my->_chain_db )
    {
-      my->_chain_db->close(true, do_flush);
+      if (after_exception)
+         my->_chain_db->close(true, my->_chain_db->undo_db_enabled());
```

Comment:
1. This code looks a bit ugly to me.
2. I haven't checked whether the whole logic is correct yet.